plugin/src/main/java/io/jenkins/plugins/redis/RedisFingerprintStorage.java

```diff
@@ -0,0 +1,168 @@
+import redis.clients.jedis.JedisPoolConfig;
+import redis.clients.jedis.exceptions.JedisException;
+
+/**
+ * Pluggable external fingerprint storage for fingerprints in Redis.
+ */
+
+@Extension
+public class RedisFingerprintStorage extends FingerprintStorage {
+
+    private final String instanceId;
+    private static final Logger logger = Logger.getLogger(Fingerprint.class.getName());
+    private static volatile JedisPool jedisPool;
+
+    public static RedisFingerprintStorage get() {
+        return ExtensionList.lookup(RedisFingerprintStorage.class).get(0);
+    }
+
+    public RedisFingerprintStorage() throws IOException {
+        try {
+            instanceId = Util.getDigestOf(new ByteArrayInputStream(InstanceIdentity.get().getPublic().getEncoded()));
+            createJedisPoolFromConfig();
+        } catch (IOException e) {
+            logger.log(Level.WARNING, "Failed to obtain Instance ID. " + e);
+            throw e;
+        }
+    }
+
+    void createJedisPoolFromConfig() {
+        GlobalRedisConfiguration config = GlobalRedisConfiguration.get();
+        createJedisPool(config.getHost(), config.getPort(), config.getConnectionTimeout(), config.getSocketTimeout(),
+                config.getUsername(), config.getPassword(), config.getDatabase(), config.getSsl());
+    }
+
+    void createJedisPool(
+            String host, int port, int connectionTimeout, int socketTimeout, String username, String password,
+            int database, boolean ssl) {
+        jedisPool = new JedisPool(new JedisPoolConfig(), host, port, connectionTimeout, socketTimeout, username,
+                password, database, "Jenkins", ssl);
+    }
+
+    private @NonNull Jedis getJedis() throws JedisException{
+        return jedisPool.getResource();
+    }
+
+    /**
+     * Saves the given fingerprint.
+     */
+    public synchronized void save(Fingerprint fp) throws JedisException {
+        Jedis jedis = null;
+        StringWriter writer = new StringWriter();
+        Fingerprint.getXStream().toXML(fp, writer);
+        try {
+            jedis = getJedis();
+            jedis.set(instanceId + fp.getHashString(), writer.toString());
+        } catch (JedisException e) {
+            logger.log(Level.WARNING, "Jedis failed in saving fingerprint: " + fp.toString() + ". " + e);
+            throw e;
+        } finally {
+            if (jedis != null) {
+                jedis.close();
+            }
+        }
+    }
+
+    /**
+     * Returns the fingerprint associated with the given unique id and the Jenkins instance ID, from the storage.
+     */
+    public @CheckForNull Fingerprint load(@NonNull String id) throws IOException, JedisException {
+        String loadedData;
+        Jedis jedis = null;
+
+        try {
+            jedis = getJedis();
+            loadedData = jedis.get(instanceId + id);
+        } catch (JedisException e) {
+            logger.log(Level.WARNING, "Jedis failed in loading fingerprint: " + id + e);
```

Comment:
Good practice: https://docs.oracle.com/javase/7/docs/api/java/util/logging/Logger.html#log(java.util.logging.Level,%20java.lang.String,%20java.lang.Throwable) for proper exception propagation in logs. Your code converts exceptions to strings, and Logging implementations cannot process exceptions after that